mount/mount_darwin.go

```diff
@@ -0,0 +1,45 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"os"
+)
+
+// Use symlink instead of union mount on darwin
+func (m *Mount) Mount(target string) error {
```

Comment:
Do we have snapshotters natively work on darwin? Should we have a new one?

Reply:
at least, make integration reported some of errors related to snapshotters (which I was struggled to interpret them).

https://github.com/containerd/containerd/runs/1064611094?check_suite_focus=true#step:10:13

It would be helpful if somebody let me know what I have to do for a proper snapshotter.

Reply:
Because bind mount is faked with a symbolic link, some features are not available.  But the current integration test for the native snapshot looks okay to me (I disabled TestSnapshotterClient/ViewReadonly for darwin); how do you think ?

```
--- PASS: TestSnapshotterClient (0.01s)
    --- PASS: TestSnapshotterClient/Chown (1.07s)
    --- PASS: TestSnapshotterClient/Basic (1.11s)
    --- PASS: TestSnapshotterClient/RemoveDirectoryInLowerLayer (1.57s)
    --- PASS: TestSnapshotterClient/Remove (1.22s)
    --- PASS: TestSnapshotterClient/Walk (1.76s)
    --- PASS: TestSnapshotterClient/PreareViewFailingtest (0.89s)
    --- PASS: TestSnapshotterClient/TransitivityTest (0.56s)
    --- PASS: TestSnapshotterClient/StatComitted (0.23s)
    --- SKIP: TestSnapshotterClient/ViewReadonly (0.00s)
    --- PASS: TestSnapshotterClient/StatActive (0.16s)
    --- PASS: TestSnapshotterClient/RootPermission (0.16s)
    --- PASS: TestSnapshotterClient/CloseTwice (0.42s)
    --- PASS: TestSnapshotterClient/Update (3.24s)
    --- PASS: TestSnapshotterClient/StatInWalk (1.11s)
    --- SKIP: TestSnapshotterClient/Rename (0.00s)
    --- PASS: TestSnapshotterClient/MoveFileFromLowerLayer (1.00s)
    --- PASS: TestSnapshotterClient/DeletedFilesInChildSnapshot (1.76s)
    --- PASS: TestSnapshotterClient/RemoveIntermediateSnapshot (1.05s)
    --- PASS: TestSnapshotterClient/DirectoryPermissionOnCommit (1.37s)
    --- PASS: TestSnapshotterClient/LayerFileupdate (8.69s)
    --- PASS: TestSnapshotterClient/128LayersMount (21.91s)
```

Reply:
This should check to make sure the mount is a bind mount before doing the symlink hack.

We should find a better snapshot solution here, what type of block device/file support does runu have here? That and the unpacking are what need to be figured out here to have something closer to a real solution.

Reply:
>This should check to make sure the mount is a bind mount before doing the symlink hack.

Okay, I will add the check before the hack.

>We should find a better snapshot solution here, what type of block device/file support does runu have here? That and the unpacking are what need to be figured out here to have something closer to a real solution.

Currently, runu supports two modes for mounting filesystems:
- a disk image file, formatted with Linux-capable filesystems (ext4, btrfs, etc)
- 9pfs hosted by server (experimental, running inside runu runtime, still a bit buggy)

So, because applications (with LKL) running over runu mount filesystem by themselves, runtimes (containerd, runu) are required to expose those files or 9pfs server ports to offer file services.

I'm not well understanding what part of _unpacking_ is not a real solution right now.

Reply:
The unpacking today is done on the host. If the host is unable to mount the snapshot filesystem, then the unpacking will need to be done inside the runtime environment. Using the native snapshotter and simulating bind mounts with symlinks is not a real solution that we would recommend to use, but gets us a test environment.

So I take it from your response that this is using the experimental 9pfs inside the runu runtime.

Reply:
Thanks, I got the point of what the native snapshotter is.

A decent snapshotter for darwin would be nice rather than the one of testing purpose, but I would expect this to further contribution because it's not likely to be a trivial component.

Or, is not having a real snapshotter a blocker for this pull request ?

> So I take it from your response that this is using the experimental 9pfs inside the runu runtime.

Yes, plus a disk image file unpacked by the snapshotter also can be mounted.

Reply:
Does the `native` snapshotter work here?  It'll be slower, but it's very simple.

Reply:
> Does the `native` snapshotter work here? It'll be slower, but it's very simple.

Yes.  This PR uses native snapshotter indeed.  On darwin there is no bind mount nor nullfs (yes it actually has but depends on fuse/bindfs etc), thus emulates with symlinks.